File: repos/safewords-ios/Safewords/Crypto/Primitives.swift
import CryptoKit
import Foundation

struct PrimitivePhrase: Equatable {
    let offset: Int
    let adjIndex: Int
    let nounIndex: Int
    let number: Int
    let phrase: String
    let hashHex: String
}

struct NumericPrimitive: Equatable {
    let counter: Int64
    let offset: Int
    let code: String
    let hashHex: String
}

struct ChallengeAnswerRow: Equatable, Identifiable {
    let rowIndex: Int
    let ask: PrimitivePhrase
    let expect: PrimitivePhrase

    var id: Int { rowIndex }
}

enum Primitives {
    static let staticOverrideLabel = "safewords/static-override/v1"

    static func staticOverride(seed: Data) -> PrimitivePhrase {
        standardWordDerivation(hash: hmac(seed: seed, message: Data(staticOverrideLabel.utf8)))
    }

    static func numeric(seed: Data, intervalSeconds: Int, timestamp: TimeInterval) -> NumericPrimitive {
        let counter = Int64(floor(timestamp / Double(intervalSeconds)))
        var bigEndian = counter.bigEndian
        let message = Data(bytes: &bigEndian, count: 8)
        let hash = hmac(seed: seed, message: message)
        let offset = Int(hash[31] & 0x0f)
        let codeInt = (
            (Int(hash[offset] & 0x7f) << 24)
                | (Int(hash[offset + 1] & 0xff) << 16)
                | (Int(hash[offset + 2] & 0xff) << 8)
                | Int(hash[offset + 3] & 0xff)
        ) % 1_000_000

        return NumericPrimitive(
            counter: counter,
            offset: offset,
            code: String(format: "%06d", codeInt),
            hashHex: hash.hexString
        )
    }

    static func challengeAnswerRow(seed: Data, tableVersion: Int = 1, rowIndex: Int) -> ChallengeAnswerRow {
        let askLabel = "safewords/challenge-answer/v\(tableVersion)/ask/\(rowIndex)"
        let expectLabel = "safewords/challenge-answer/v\(tableVersion)/expect/\(rowIndex)"
        return ChallengeAnswerRow(
            rowIndex: rowIndex,
            ask: standardWordDerivation(hash: hmac(seed: seed, message: Data(askLabel.utf8))),
            expect: standardWordDerivation(hash: hmac(seed: seed, message: Data(expectLabel.utf8)))
        )
    }

    static func challengeAnswerRows(seed: Data, tableVersion: Int = 1, count: Int) -> [ChallengeAnswerRow] {
        (0..<count).map { challengeAnswerRow(seed: seed, tableVersion: tableVersion, rowIndex: $0) }
    }

    static func standardWordDerivation(hash: [UInt8]) -> PrimitivePhrase {
        let offset = Int(hash[31] & 0x0f)
        let adjIndex = ((Int(hash[offset] & 0x7f) << 8) | Int(hash[offset + 1] & 0xff)) % TOTPDerivation.adjectiveCount
        let nounIndex = ((Int(hash[offset + 2] & 0x7f) << 8) | Int(hash[offset + 3] & 0xff)) % TOTPDerivation.nounCount
        let number = ((Int(hash[offset + 4] & 0x7f) << 8) | Int(hash[offset + 5] & 0xff)) % TOTPDerivation.numberMod
        let phrase = "\(TOTPDerivation.adjectives[adjIndex]) \(TOTPDerivation.nouns[nounIndex]) \(number)"
        return PrimitivePhrase(
            offset: offset,
            adjIndex: adjIndex,
            nounIndex: nounIndex,
            number: number,
            phrase: phrase,
            hashHex: hash.hexString
        )
    }

    private static func hmac(seed: Data, message: Data) -> [UInt8] {
        let key = SymmetricKey(data: seed)
        return Array(HMAC<SHA256>.authenticationCode(for: message, using: key))
    }
}

private extension Array where Element == UInt8 {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}

File: repos/safewords-android/app/src/main/kotlin/com/thc/safewords/crypto/Primitives.kt
package com.thc.safewords.crypto

import com.thc.safewords.data.WordLists
import java.nio.ByteBuffer
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Primitives {

    private const val ADJECTIVE_COUNT = 197
    private const val NOUN_COUNT = 300
    private const val NUMBER_MODULUS = 100

    private const val OVERRIDE_LABEL = "safewords/static-override/v1"

    fun staticOverride(seed: ByteArray): String {
        require(seed.size == 32) { "Seed must be 32 bytes" }
        val hash = hmacSha256(seed, OVERRIDE_LABEL.toByteArray(Charsets.UTF_8))
        return phraseFromHash(hash)
    }

    fun numeric(seed: ByteArray, intervalSeconds: Int, timestamp: Long): String {
        require(seed.size == 32) { "Seed must be 32 bytes" }
        require(intervalSeconds > 0) { "Interval must be positive" }
        val counter = timestamp / intervalSeconds
        val counterBytes = ByteBuffer.allocate(8).putLong(counter).array()
        val hash = hmacSha256(seed, counterBytes)
        val offset = hash[31].toInt() and 0x0F
        val codeInt = (
            ((hash[offset].toInt() and 0x7F) shl 24) or
                ((hash[offset + 1].toInt() and 0xFF) shl 16) or
                ((hash[offset + 2].toInt() and 0xFF) shl 8) or
                (hash[offset + 3].toInt() and 0xFF)
            ) % 1_000_000
        return "%06d".format(codeInt)
    }

    data class ChallengeAnswerRow(val rowIndex: Int, val ask: String, val expect: String)

    fun challengeAnswerRow(seed: ByteArray, tableVersion: Int, rowIndex: Int): ChallengeAnswerRow {
        require(seed.size == 32) { "Seed must be 32 bytes" }
        val askLabel = "safewords/challenge-answer/v$tableVersion/ask/$rowIndex"
        val expectLabel = "safewords/challenge-answer/v$tableVersion/expect/$rowIndex"
        val askHash = hmacSha256(seed, askLabel.toByteArray(Charsets.UTF_8))
        val expectHash = hmacSha256(seed, expectLabel.toByteArray(Charsets.UTF_8))
        return ChallengeAnswerRow(
            rowIndex = rowIndex,
            ask = phraseFromHash(askHash),
            expect = phraseFromHash(expectHash),
        )
    }

    fun challengeAnswerTable(seed: ByteArray, tableVersion: Int, rowCount: Int): List<ChallengeAnswerRow> =
        (0 until rowCount).map { challengeAnswerRow(seed, tableVersion, it) }

    private fun phraseFromHash(hash: ByteArray): String {
        val offset = hash[31].toInt() and 0x0F
        val adjIdx = ((hash[offset].toInt() and 0x7F) shl 8 or
            (hash[offset + 1].toInt() and 0xFF)) % ADJECTIVE_COUNT
        val nounIdx = ((hash[offset + 2].toInt() and 0x7F) shl 8 or
            (hash[offset + 3].toInt() and 0xFF)) % NOUN_COUNT
        val number = ((hash[offset + 4].toInt() and 0x7F) shl 8 or
            (hash[offset + 5].toInt() and 0xFF)) % NUMBER_MODULUS
        return "${WordLists.adjectives[adjIdx]} ${WordLists.nouns[nounIdx]} $number"
    }

    private fun hmacSha256(key: ByteArray, message: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(message)
    }
}
